File: Codeforces/1400/1443C/main.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#define ll long long
using namespace std;

int main()
{
    ios_base::sync_with_stdio(false);
    cin.tie(NULL);
    int tc;
    cin >> tc;
    while (tc--)
    {
        int n;
        cin >> n;
        vector<int> delivery(n);
        vector<int> pick_up(n);
        vector<int> idx;

        ll left = 0;
        ll right = 0;

        for (int i = 0; i < n; i++)
            cin >> delivery[i];
        for (int i = 0; i < n; i++)
        {
            cin >> pick_up[i];
            if (delivery[i] > pick_up[i])
            {
                idx.push_back(i);
            }
            else
            {
                left = max(left, (ll)delivery[i]);
            }
            right += pick_up[i];
        }

        ll ans = right;
        while (left <= right)
        {
            ll mid = (left + right) / 2;
            ll sum = 0;
            for (int i = 0; i < idx.size(); i++)
            {
                if (delivery[idx[i]] > mid)
                {
                    sum += pick_up[idx[i]];
                }
            }
            if (sum <= mid)
            {
                ans = min(ans, mid);
                right = mid - 1;
            }
            else
            {
                left = mid + 1;
            }
        }

        cout << ans << '\n';
    }
}

File: Codeforces/1400/1443C/main.java
import java.io.*;
import java.util.*;

public class Main {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));

        int tc = Integer.parseInt(br.readLine());
        while (tc-- != 0) {
            int n = Integer.parseInt(br.readLine());

            int[] delivery = new int[n];
            int[] pickUp = new int[n];

            StringTokenizer st = new StringTokenizer(br.readLine());
            for (int i = 0; st.hasMoreTokens(); i++) {
                delivery[i] = Integer.parseInt(st.nextToken());
            }

            st = new StringTokenizer(br.readLine());
            for (int i = 0; st.hasMoreTokens(); i++) {
                pickUp[i] = Integer.parseInt(st.nextToken());
            }

            long left = 0;
            long right = 0;
            List<Integer> idx = new ArrayList<>();
            for (int i = 0; i < n; i++) {
                if (delivery[i] > pickUp[i]) {
                    idx.add(i);
                } else {
                    left = Math.max(left, delivery[i]);
                }
                right += pickUp[i];
            }

            long ans = right;
            while (left <= right) {
                long mid = (left + right) / 2;
                long sum = 0;
                for (Integer i : idx) {
                    if (delivery[i] > mid) {
                        sum += pickUp[i];
                    }
                }
                if (sum > mid) {
                    left = mid + 1;
                } else {
                    ans = Math.min(ans, mid);
                    right = mid - 1;
                }
            }

            bw.write("" + ans + "\n");
        }
        bw.flush();
        bw.close();
        br.close();
    }
}